# Task Sheets

Biweekly Task Sheet Submissions - Machine Learning and Statistics - Winter 2023 / 2024

Lecturer Ian McLoughlin

Task Sheet PDFs included in Supporting Data Folder

***


Tasks

[Newtons Method](#taskone)

[Chi-Squared](#tasktwo)

[T-Test](#taskthree)

[TBC](#taskfour)

[TBC](#taskfive)



****

Imports

In [14]:
# Using Python Packages this can also be achieve by the following:

# maths
import math

# numerical arrays
import numpy as np

# dataframes
import pandas as pd

# from Scipy Stats - Contingency Chi Squared
from scipy.stats import chi2_contingency

<a id="taskone"></a> 

## Newtons Method

> Write a function $ sqrt (x) $ to approximate the square root of a floating point number $ x $ without using the power operator or a package.

> Rather you should use Newton's method. Start with an initial guess for the square root called $ z_{0} $. You then repeatly improve it using the folowing formula, until the difference between a previous guess $ z_{i} $ and the next $ z_{i + 1} $ is less than some threshold for example 0.01.


$$
  z_{i + 1}  =  z_{i} - \frac{ \ z_{i}  *  z_{i} - x }{ 2z_{i} }
$$


<br/>


T1 Q1 Links:

[Go Programming language](https://go.dev/tour/welcome/1)

[Loops and Functions in Go](https://go.dev/tour/flowcontrol/8)

[Newton Sq Rt](https://math.mit.edu/~stevenj/18.335/newton-sqrt.pdf)


[1][2][3][4][5]

To test our function we will use test data - we will use data where we know the outcome and test that the program finds this answer.

A Square root of a number is a value that when multipled by itself will give the original number. [6]

Newtons Function asks us to start with a guess point and when the function is run repeatedly we get the result. As outlined in UT Calculus; "In typical situations, Newton's method homes in on the answer extremely quickly, roughly doubling the number of decimal points in each round. So if your original guess is good to one decimal place, 5 rounds later you will have an answer good to 30+ digits." [7]

Test Variables: To test the function we will look for the square root of 16, which we know to be 4. As as starting point we will guess 2 to be the correct answer. Using these variables / parameters we will run the function.

Once we have confirmed that our function is working correctly and we have tested it we can then change variables and use it to calculate the square root of other numbers.

In [2]:
# Alternative Options to Newtons Method:

# python will allow you to work out the Square Root as following -

16 ** 0.5


# in this instance we are raising 16 to the power of 0.5 

# [8]

4.0

In [3]:
# square root of the number 16
math.sqrt(16) 

# for this page you cannot use a negative number - these will always be positive numbers
# however for this Question we are using Newtons Method as opposed to power operators or packages.

# [9]

4.0

We will now plot the Equation - inputting our variables

In [4]:
# Set x  - the number we want to calculate the square root of

x = 16

In [5]:
# Set z - this number is our initial guess or starting point for what we think the square root could be

z = 2.0

# note we are setting this as a floating number
# unless the Square root is a full number, or perfect square, outcomes will be a floating point

In [6]:
# First Approximation


znext = z - (((z * z) - x) / ( 2*z )) 

# setting a variable called znext - or our next guess
# using Newtons formula above write a formula in Python
# note 2z is written as 2*z in Python
# precendence operators - ensures that multiplication before subtraction

# znext equals, z minus the results of z multiplied by z minus x, divided by 2z (z * 2)

znext

# print znext

5.0

Evaluation of first run:

Is 5.0 a better approximation?

Yes, 2 is 2 units away from 4, where as 5 is 1 unit away.

Continuing to run this formula and you will get you closer to the required number 4. This shows the repeating nature of this formula and its importance in Machine Learning - this formula will more accurate with each time it is run further writing full functions will allow for us to set it to run with a error margain and also a maximum number of iterations.

In [7]:
# Write a formula to complete this process in one instance
# the following is a version of Newton's method where we set the number of times where the funcion will run - 100 times

def sqrt(x):
    # initial guess for the sq rt
    z = x / 4.0
    # iteration is used in this function - this function is run 100 times - this can be set as a variable and run any number of times
    for i in range(100):
        # Newton's method for a better approximation.
        z - (((z * z) - x) / ( 2*z )) 


    return z


In [8]:
# run the above function

sqrt(16)

4.0

In the above method we are using iteraiton and setting the number of times the function will run, in this instance we have set this for 100. What if we have a  number where to get the correct value we need to run the function 101 times?

There for a while loop which would complete evaluations and continue to run while conditions are being met is a better version.

In [9]:
# here we use a while loop to continue to fun the formula while conditions are being met
# We use both an approximate as a starting point and then incoropating the better guess from each time the code is run

def while_sqrt(n):
    approx=0.5*n
    better=0.5*(approx+n/approx)
    while better!=approx:
        approx=better
        better=0.5*(approx+n/approx)
    return approx



# [10]

In [10]:
# run the above function on our test number 16

while_sqrt(16)

4.0

In [11]:
while_sqrt(2)

1.414213562373095

![alternative text](Images/Newton-Raphson-Method.jpg)


[11]

## Additional Notes


1. Limitations to this method are as follows "the method is not efficient to calculate the roots of the polynomials or equations with higher degrees but in the case of small-degree equations, this method yields very quick results." [11]


2. A alternative way to think of this method is that it is a way to find 0 (root) or the X intercept of a function:

$$ f(x) = 0   -->  f(x) = 0 x^2 --> x^2 = 2 $$

[12][13]

3. Convergence rates can and will differ depending on our initial guess, convergences mean that we are approaching the answer [14] This is important for any implementation - both to know if it will be successful and also the time it takes to reach the correct answer. 

Example of convergence rates:

Here we are looking for the square root of 2 - in this graph you will see the the function in blue and the derivative of this function in yellow:

![alternative text](Images/Newton_root2_graph.jpg)

<br/>

Here we can see indicated both fast and slower convergence areas - high dots indication areas of slow convergence and low dots where there are areas of fast convengence. This shows the importance of the first guess and its impact on outcomes in terms of speed and relability.

![alternative text](Images/Newtons_Convergence_root2.jpg)

[12]

Note:
"The derivative of a function describes the function's instantaneous rate of change at a certain point. Another common interpretation is that the derivative gives us the slope of the line tangent to the function's graph at that point." [15]


<br/>

Real World Uses??????????????



***
<a id="tasktwo"></a> 


## Chi-squared






> Consider a contingency table, based on a survey asking preferences between coffee or tea, and plain & chocolate biscuits.

> Use scipy.stats to perform a chi-squared test to see whether there is any evidence of an association between drink preference and biscuit preference.


              

|             | *biscuits* Chocolate  | Plain   |
| -:| -: | -: |
| *Drink* Coffee|        43  |      57 |
|       Tea   |        56  |      45 |




<br/>

What is a Chi=Square Test?

"The Chi-Square test is a statistical procedure for determining the difference between observed and expected data."

"This test can also be used to determine whether it (the difference) correlates to the categorical variables in our data. It helps to find out whether a difference between two categorical variables is due to chance or a relationship between them." [15]

Formula for chi-square:

$$

  X ^2 = \sum \frac {  (O_{i} - E_{i})^2} { E_{i} }
  
$$



$ X^2 $ = Chi-square value

$ O_{i} $ = Observed Value

$ E_{i} $ = Expected Value

[16]

<br/>


$$

  H_{0} 
  
$$

The Null Hypothesis

The null hypothesis states that there is no association between the variables and therefore the variables are independent statisically speaking.

When testing data, we start from a point of null - we start by thinking that there will be no correlation between the variables / numbers and by examining the data as a whole we hope to prove this null hypotesis incorrect, therefore proving that there is a relationship present.

We use the Chi-Square test to compare the variables and the correlation between them.

[17]

<br/>



Types of Chi-Squared Test

The main types of chi-squared tests are Goodness to Fit tests and test of Independence.

Goodness to Fit : " The Chi-square goodness of fit test is a statistical hypothesis test used to determine whether a variable is likely to come from a specified distribution or not. It is often used to evaluate whether sample data is representative of the full population. " [18]

Test of Independance : "The Chi-square test of independence is a statistical hypothesis test used to determine whether two categorical or nominal variables are likely to be related or not." [19]




Comparison of Chi-Square Tests


![alternative text](Images/Chi_sq_comparison.jpg)


[20]
<br/>

Importance of contingency Table and understanding our Data

<br/>

"A contingency table is a tabular representation of categorical data . A contingency table usually shows frequencies for particular combinations of values of two discrete random variable s X and Y. Each cell in the table represents a mutually exclusive combination of X-Y values." [21]

A discrete random variable is one where a count has been obtained; where there is a countable value for X, for example the number of people who's favourite colour is red in a given population, the number of people at a concert.

This allows us to ask definate questions and take definate measurements.

[21][22]

<br/>

In the below table I have added in the totals which allows us to get a better picture of the data as a whole:



|             | *biscuits* Chocolate  | Plain   | <span style="color:red">Total </span>
| -:| -: | -: |-: |
| *Drink* Coffee|        43  |      57 | <span style="color:red">    100 </span>|
|       Tea   |        56  |      45 | <span style="color:red">    101 </span>|
| <span style="color:red">    Total </span>  | <span style="color:red">        99  </span>| <span style="color:red">     102 </span>|  <span style="color:red">   201 </span>|


For this we can see we have taken a total sample of 99 People, who are Chocolate Biscuit eaters who drink either tea or coffee. There is a total number of 102 People who drink either tea or coffee and prefer plain biscuits. 

The total number of People who drink coffee are 100, and the total who drink tea are 101.

A larger number of tea drinkers prefer chocolate biscuits while a larger number of coffee drinker prefer plain biscuits.

Completing a chi-squared test will give us a better understanding of this data and if there is a correlation between the groups.

Potential hypotesis:

$ H_{0} $ ( First Hypothises) - That there is no correlation between the drink type and biscuit type

$ H_{1} $ ( Second Hypothises) - That there is a correlation between the drink type and biscuit type



[23]

In [12]:

table = np.array([[43, 57],
                  [56, 45]])

chi2, p, dof, expected = chi2_contingency(table)

if (p <= 0.05):
    print("Reject the Null Hypotesis")
else:
    print("Fail to reject Null Hypothesis")

print(f"Statistical Value:     {chi2:.5g}")
print(f"p-value:            {p:.5g}")
print(f"degrees of freedom: {dof}")
print("expected frequencies:")
print(expected)



# [24]

Fail to reject Null Hypothesis
Statistical Value:     2.6359
p-value:            0.10447
degrees of freedom: 1
expected frequencies:
[[49.25373134 50.74626866]
 [49.74626866 51.25373134]]


When running scipy.stats.chi2_contingency in python the outputs will be; statistic (float), pvalue(float), dof (integer), experected_freq (numpy array - outputed to match shape of inputted data). [25]

<br/>

Terms explained:

- significance level: This is the measure of the strength of the evidence that must be present before you can reject the null hypothesis, therefore before you can start that there is no correlation. [17] "The significance level is a predetermined probability threshold - commonly 0.05" [26]

- Statistical Value:  

- p-value: "quantifies how strongly the sample data contradict the null hypothesis", you can reject the null hypotesis by comparing the p-value to the signifcance level, where the pvalue is less than the significance level you can accept the null hypotesis. [17]

![alternative text](Images/p_Value.jpg)

[27]


- degrees of freedom:

- expected frequencies:

note: A low p-value is evidence against the null hypothesis. In this case, the null hypothesis is that there is no association among the categories. A large discrepancy of the frequencies from the expected frequencies suggests that there is some sort of association or dependence. Roughly: large discrepancy ⇒ large chi-square statistic ⇒ small p-value.


What is a low P Value?!

|             | *biscuits* Chocolate  | Plain   | <span style="color:red">Total </span>
| -:| -: | -: |-: |
| *Drink* Coffee|        43  |      57 | <span style="color:red">    100 </span>|
| Expected|        49.3  |      50.7 | <span style="color:green">    49.7% </span>|
|       Tea   |        56  |      45 | <span style="color:red">    101 </span>|
| Expected|        49.7 |      51.3 | <span style="color:green">    50.3% </span>|
| <span style="color:red">    Total </span>  | <span style="color:red">        99  </span>| <span style="color:red">     102 </span>|  <span style="color:red">   201 </span>|

Conversely, when the p-value is greater than your significance level, you fail to reject the null hypothesis. The sample data provides insufficient data to conclude that the effect exists in the population.  (https://statisticsbyjim.com/hypothesis-testing/null-hypothesis/)

In [13]:

table = np.array([[90, 60, 104, 95],
                  [30, 50,  51, 20],
                  [30, 40,  45, 35]])

chi2, p, dof, expected = chi2_contingency(table)

if (p <= 0.05):
    print("Reject the Null Hypotesis")
else:
    print("Fail to reject Null Hypotesis")

print(f"chi2 statistic:     {chi2:.5g}")
print(f"p-value:            {p:.5g}")
print(f"degrees of freedom: {dof}")
print("expected frequencies:")
print(expected)

Reject the Null Hypotesis
chi2 statistic:     24.571
p-value:            0.00040984
degrees of freedom: 6
expected frequencies:
[[ 80.53846154  80.53846154 107.38461538  80.53846154]
 [ 34.84615385  34.84615385  46.46153846  34.84615385]
 [ 34.61538462  34.61538462  46.15384615  34.61538462]]


***

<br/>

<a id="taskthree"></a>

## T-Test


> Perform a t-test on the famous penguins data set

> Investigate whether there is evidence of signicant difference in the body and mass of male and female gentoo penguins

links:


First we will import the dataset using Pandas. We will look at the dataset to get a better idea of the data included.

In [15]:
# import dataset
df = pd.read_csv('DataSets/penguins.csv')

df

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
3    Adelie  Torgersen             NaN            NaN                NaN   
4    Adelie  Torgersen            36.7           19.3              193.0   
..      ...        ...             ...            ...                ...   
339  Gentoo     Biscoe             NaN            NaN                NaN   
340  Gentoo     Biscoe            46.8           14.3              215.0   
341  Gentoo     Biscoe            50.4           15.7              222.0   
342  Gentoo     Biscoe            45.2           14.8              212.0   
343  Gentoo     Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  
0         3750.0    MALE  
1         3800.0  FEMALE  
2         3250.0  FEMALE  
3            NaN     NaN  
4         3450.0  FEMALE  
..           ...     ...  
339          NaN     NaN  
340       4850.0  FEMALE  
341       5750.0    MALE  
342       5200.0  FEMALE  
343       5400.0    MALE  

[344 rows x 7 columns]

In [34]:
# get the data info - this is a general overview of the dataframe

df.info()

# we can see from this that there are between 344 - 333 entries so we have null values included
# we can also see that we have two different data types - objects and floats = names and numbers


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    object 
dtypes: float64(4), object(3)
memory usage: 18.9+ KB


In [32]:
df.head(10)

# take the top ten rows and look at the df
# from this we can see that there are entries with no values across the entire row


species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   
3  Adelie  Torgersen             NaN            NaN                NaN   
4  Adelie  Torgersen            36.7           19.3              193.0   
5  Adelie  Torgersen            39.3           20.6              190.0   
6  Adelie  Torgersen            38.9           17.8              181.0   
7  Adelie  Torgersen            39.2           19.6              195.0   
8  Adelie  Torgersen            34.1           18.1              193.0   
9  Adelie  Torgersen            42.0           20.2              190.0   

   body_mass_g     sex  
0       3750.0    MALE  
1       3800.0  FEMALE  
2       3250.0  FEMALE  
3          NaN     NaN  
4       3450.0  FEMALE  
5       3650.0    MALE  
6       3625.0  FEMALE  
7       4675.0    MALE  
8       3475.0     NaN  
9       4250.0     NaN

In [33]:
# breakdown the file by species
species = df.groupby("species").size() 

species

# three species are listed


species
Adelie       152
Chinstrap     68
Gentoo       124
dtype: int64

In [26]:
# breakdown the file by Island
islands = df.groupby("island").size() 

islands
# three islands are listed

island
Biscoe       168
Dream        124
Torgersen     52
dtype: int64

In [36]:
# breakdown the file by Sex
sexes= df.groupby("sex").size() 

sexes
# we have two sexes listed and the n/a values are not included in this output
# the total number shown here is 333


sex
FEMALE    165
MALE      168
dtype: int64

In [39]:
# find out where the missing values are

missing_val = df.isin([pd.NA, np.nan, None, '-', 'N/A', 0]).sum()

missing_val

# https://stackoverflow.com/questions/74305111/how-to-count-missing-values-that-are-not-the-regular-missing-in-pandas

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
dtype: int64

In [44]:
# now take a subset of the Gentoo Pengiun Group

gentoo = df.loc[df["species"] =="Gentoo"] 

gentoo


species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
220  Gentoo  Biscoe            46.1           13.2              211.0   
221  Gentoo  Biscoe            50.0           16.3              230.0   
222  Gentoo  Biscoe            48.7           14.1              210.0   
223  Gentoo  Biscoe            50.0           15.2              218.0   
224  Gentoo  Biscoe            47.6           14.5              215.0   
..      ...     ...             ...            ...                ...   
339  Gentoo  Biscoe             NaN            NaN                NaN   
340  Gentoo  Biscoe            46.8           14.3              215.0   
341  Gentoo  Biscoe            50.4           15.7              222.0   
342  Gentoo  Biscoe            45.2           14.8              212.0   
343  Gentoo  Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  
220       4500.0  FEMALE  
221       5700.0    MALE  
222       4450.0  FEMALE  
223       5700.0    MALE  
224       5400.0    MALE  
..           ...     ...  
339          NaN     NaN  
340       4850.0  FEMALE  
341       5750.0    MALE  
342       5200.0  FEMALE  
343       5400.0    MALE  

[124 rows x 7 columns]

In [45]:
# check number of rows to see if they match with the earlier count of 124

number_of_gentoo = len(gentoo)

number_of_gentoo

124

In [49]:
# check where our missing values are

missing_val_gentoo = gentoo.isin([pd.NA, np.nan, None, '-', 'N/A', 0]).sum()

missing_val_gentoo

species              0
island               0
bill_length_mm       1
bill_depth_mm        1
flipper_length_mm    1
body_mass_g          1
sex                  5
dtype: int64

In [50]:
# check the breakdown of the sexes

gentoo_sexes= gentoo.groupby("sex").size() 

gentoo_sexes

# we are missing 5 as above
# 58 + 61 = 119

# total entries are 124 
# 124 - 5 = 119

sex
FEMALE    58
MALE      61
dtype: int64

We have now a better understanding of the overall dataframe, and the Gentoo subset. We can see what is listed and also where the null values are.

As we need to look at the body mass of male and female - however we know we are missing 5 entries.

Lets look closer at these entries and see if there are body mass values listed

In [ ]:
#gentoo = df.loc[df["species"] =="Gentoo"] 

missing_gentoo = gentoo.loc[gentoo["sex"] == ""]

# Bibliography

Task 1
Newtons Method

[1] Linuxhint, How to use latex in Jupyter Notebook, 2021, https://linuxhint.com/use-latex-jupyter-notebook/

[2] Patrick Walls Github, 2022 https://patrickwalls.github.io/mathematicalpython/jupyter/latex/ & https://github.com/patrickwalls/mathematicalpython/blob/master/notebooks/jupyter/latex.ipynb

[3] StackOverflow, Centering Text in IPython notebook markdown/heading cells?, 2013 https://stackoverflow.com/questions/18575012/centering-text-in-ipython-notebook-markdown-heading-cells

[4] Maths and Equations, Jupyterbook.org, 2022, https://jupyterbook.org/en/stable/content/math.html

[5] MyST Markedly Structured Text, Sept 20 2023, https://myst-parser.readthedocs.io/en/latest/syntax/optional.html#syntax-amsmath 

[6] Splash Learn, What Are Square and Square Roots?, https://www.splashlearn.com/math-vocabulary/algebra/square-and-square-roots#:~:text=The%20square%20root%20of%20a,themselves%2C%20give%20the%20product%2025.

[7] UT Calculus, Department of Mathematics at UT Austin, https://web.ma.utexas.edu/users/m408n/CurrentWeb/LM4-8-2.php#:~:text=Newton's%20method%20is%20a%20technique,a%20new%20(and%20improved!)

[8] Enterprise DNA Blog, Python ** Explained (Double Star or Double Asterix),  https://blog.enterprisedna.co/what-does-double-star-mean-in-python/#:~:text=The%20double%20star%20operator%2C%20denoted,calls%20to%20unpack%20keyword%20arguments.

[9] Real Python, The Python Square Root Function, July 08 2019, https://realpython.com/python-square-root-function/

[10] Medium.com, Python program to find square root of the number using Newton’s method, 2021 https://thirumalai2024.medium.com/
python-program-to-find-square-root-of-the-number-using-newtons-method-937c0e732756 

[11] Geeks for Geeks, Newton Raphson Method, 4th July 2023  https://www.geeksforgeeks.org/newton-raphson-method/ 

[12] Computing Skills Set, 11 Highly Instructive Examples for the Newton Raphson Method, https://computingskillset.com/solving-equations/highly-instructive-examples-for-the-newton-raphson-method/ 

[13] Newton's Method, PatrickMJT, Youtube 2013, https://www.youtube.com/watch?v=1uN8cBGVpfs 

[13] Kahn Academy, Convergent and divergent sequences https://www.khanacademy.org/math/calculus-2/cs2-series/cs2-convergent-and-divergent-infinite-series/v/convergent-and-divergent-sequences 

[14] Kahn Academy, Derivatives: definition and basic rules, https://www.khanacademy.org/math/differential-calculus/dc-diff-intro#:~:text=The%20derivative%20of%20a%20function,function's%20graph%20at%20that%20point.


Chi-Squared

[15] SimpliLearn, What is a Chi-Square Test? Formula, Examples & Application, Oct 11th 2023, https://www.simplilearn.com/tutorials/statistics-tutorial/chi-square-test 

[16] CueMath, Chi Square Formula,  https://www.cuemath.com/chi-square-formula/

[17] Statistics by Jim, Null Hypothesis: Definition, Rejecting & Examples, 2023 https://statisticsbyjim.com/hypothesis-testing/null-hypothesis/ 

[18] JMP, Chi-Square Goodness of Fit Test, 2023, https://www.jmp.com/en_be/statistics-knowledge-portal/chi-square-test/
chi-square-goodness-of-fit-test.html 

[19] JMP, Chi-Square Test of Independence, 2023  https://www.jmp.com/en_be/statistics-knowledge-portal/chi-square-test/chi-square-test-of-independence.html 

[20] JMP, The Chi-Square Test, 2023, https://www.jmp.com/en_be/statistics-knowledge-portal/chi-square-test.html 

[21] Statistics.com, Contingency Table, https://www.statistics.com/glossary/contingency-table/#:~:text=A%20contingency%20table%20is%20a,exclusive%20combination%20of%20X%2DY%20values. 

[22] Henry County Schools, Discrete and Continuous Random Variableshttps://www.henry.k12.ga.us/ugh/apstat/chapternotes/7supplement.html#:~:text=A%20discrete%20variable%20is%20a,red%20marbles%20in%20a%20jar 

[23] Khan Academy,  Contingency table chi-square test, https://www.khanacademy.org/math/statistics-probability/inference-categorical-data-chi-square-tests/chi-square-tests-for-homogeneity-and-association-independence/v/contingency-table-chi-square-test 

[24] StackOverflow, understanding scipy.stats.chisquare, 2021 https://stackoverflow.com/questions/64669448/understanding-scipy-stats-chisquare 

[25] scipy.org,  scipy.stats.chi2_contingency, 2008- 2023 https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html 

[26] Visual Design, An Interactive Guide to Hypothesis Testing in Python, jun 12th, 2022 https://www.visual-design.net/post/an-interactive-guide-to-hypothesis-testing-in-python#:~:text=The%20significance%20level%20is%20a,the%20null%20hypothesis%20is%20true. 

[27] CueMath, Chi Square Formula,  https://www.cuemath.com/chi-square-formula/ 


t-test

[]
